### PR TITLE
IPv4|IPv6|host with optional port pair validator added.

### LINKED ISSUE #3: Support `host:port` validator
### Description
For Logstash-to-Logstash communication, we currently use HTTP under the hood but it is an implementation detail. That means, in the future it subjects to any other protocol change. Thus, we will align on <host:port> pair and the validation should be included in this mixin package.

### PR DESCRIPTION
**Pull request by mashhurs**
### What does this PR do?
It adds a validator for <host:port> pair where port is optional, default port can be applied.

- Closes #3

### CODE REVIEW

File: spec/logstash/plugin_mixins/validator_support/host_port_pair_validation_adapter_spec.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    [
+      ["127.0.0.1:1234"],
+      ["82.31.1.3:9800"],
+      ["foo.com:1234"],
+      ["foo-bar-domain.com:9800"],
+      ["::1:9800"],
+      ["::1"],
+      ["2001:db8:3333:4444:5555:6666:7777:8888:9800"],
+      ["::ffff:93.184.216.34:1023"],
+      ["127.0.0.1"],
```
Comment (mashhurs) [inline spec/logstash/plugin_mixins/validator_support/host_port_pair_validation_adapter_spec.rb]: review note: port is optional where default port can be applied.

Comment (yaauie) [reply spec/logstash/plugin_mixins/validator_support/host_port_pair_validation_adapter_spec.rb]: >  port is optional where default port can be applied.

Then I would name this validator `required_host_optional_port_pair`

File: lib/logstash/plugin_mixins/validator_support/host_port_pair_validation_adapter.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# encoding: utf-8
+
+require 'logstash/plugin_mixins/validator_support'
+
+module LogStash
+  module PluginMixins
+    module ValidatorSupport
+
+      host_port_regex = %r{(?:(?:[a-z][a-z0-9\-\.])|(([0-9a-fA-F]{0,4}:){1,7}[0-9a-fA-F]{0,4})|(?:\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b))(:?:[0-9]{1,5})?$}xi
```
Comment (mashhurs) [inline lib/logstash/plugin_mixins/validator_support/host_port_pair_validation_adapter.rb]: Review note: used online regex tool for more tests: https://regex101.com/r/c0Vrac/1

**Comment by yaauie** [comment]
I think we need to synchronize on the requirements. In the issue comment that spawned the need to do this work, I said:

> We should accept _only_ colon-separated hostname-port pairs, formally:
> 
> ```antlr
> hostsEntry: host ':' port
>           ;
> 
> host: ipAddress
>     | hostname
>     ;
> 
> ipAddress:  ipv4Address
>           | '[' ipv6Address ']'
>           ;
> 
> # ...
> # host is a valid domain name (e.g. `myserver1.example.com`)
> # ipv4Address is dot-decimal representation (e.g. `127.0.0.1`)
> # ipv6Address is any RFC4291§2.2 text representation of an ipv6 address (e.g., `::1`, `2001:0db8:85a3:0000:0000:8a2e:0370:7334`, `::ffff:93.184.216.34`)
> ```
> -- [Upstream spec](https://github.com/logstash-plugins/logstash-integration-logstash/pull/13#discussion_r1340505219)

As-implemented, we are missing the specified square brackets enclosing an IPv6 address (which allow us to differentiate between a colon separating IPv6 components and one separating the host from the port). 

I also see that you want to make the port optional, which makes sense because the downstream consumer of this new validator may provide a default port. In the context of the required square brackets around IPV6 addresses, this makes specifying a bare ipv6 address with the original requirements a little awkward.

Therefore, I think we should modify the spec in a couple of ways:

1. Name it in such a way that clearly conveys that the host is required and the port is optional (i.e., `required_host_optional_port`)
2. Add an IPv6-without-port path that does not require the square brackets, but will properly wrap bare IPv6 addresses in square brackets when building the URL in the HTTP implementation

```antlr
hostsEntry: hostPortPair
          | encodedHost
          # | ipv6Address EDIT: NO unwrapped IPv6 
          ;

hostPortPair: encodedHost ':' port

encodedHost: host
           | ipv4Address
           | '[' ipv6Address ']'
           ;

# ...
# port is a valid port number
# host is a valid domain name (e.g. `myserver1.example.com`)
# ipv4Address is dot-decimal representation (e.g. `127.0.0.1`)
# ipv6Address is any RFC4291§2.2 text representation of an ipv6 address (e.g., `::1`, `2001:0db8:85a3:0000:0000:8a2e:0370:7334`, `::ffff:93.184.216.34`)
```

Examples:

- Unchanged from original spec:
  - `example.com` -> host: `example.com`, port: _DEFAULT_
  - `example.com:1234` -> host: `example.com`, port: `1234`
  - `[::1]:1234` -> host: `::1`, port `1234`

- New from making port optional:
  - `127.0.0.1:1234` -> host: `127.0.0.1, port: `1234`
  - `127.0.0.1` -> host: `127.0.0.1`, port: _DEFAULT_
  - `[::1]` -> host: `::1`, port _DEFAULT_ 

- ~New from making no-port-IPv6 less awkward:~
  - ~`::1` -> host: `::1`, port _DEFAULT_~

**Comment by mashhurs** [comment]
> I think we need to synchronize on the requirements. In the issue comment that spawned the need to do this work, I said:
> 
...

As we talked offline, out final updated assumption:
- `ipv6Address` requires enclosed bracket
- port range aligned with 0~65535

**Review by yaauie — COMMENTED** [review]
I think this looks right, but would prefer to have #5 in place and a green run with modern CI targets before merging this.

**Comment by mashhurs** [comment]
> I think this looks right, but would prefer to have #5 in place and a green run with modern CI targets before merging this.

Reviewed, merged #5 and rebased #4.

**Review by yaauie — APPROVED** [review]
LGTM!

**Pull request merged**